app/services/re_captcha.rb

```diff
@@ -0,0 +1,52 @@
+# This service encapsulates some logic related to reCAPTCHA.
+#
+# The `enabled?` and `disabled?` methods will tell if the reCAPTCHA is
+# necessary (enabled) or if it's not necessary (disabled) in the current
+# context. This is determined by the `current_user`, or the lack thereof.
+#
+# Example: ReCaptcha.call(current_user).enabled? => true/false
+class ReCaptcha
+  include Devise::Controllers::Helpers
+
+  def self.call(current_user = nil)
+    new(current_user).call
+  end
+
+  def initialize(current_user)
```

Comment:
Is supporting both singleton and instance invocation a conscious choice? I'm curious how we feel about that: I like the flexibility, but wonder if it's worth the added complexity?

Reply:
It is a conscious choice in the sense that using the class methods help shorten the syntax for the registration form. This is only because it's known/expected to never need to pass in a `current_user` (because it's a registration feature).

```ruby
ReCaptcha.for_registration_disabled? # current approach

# vs

ReCaptcha.call.for_registration_disabled?
```

In all other situations we do need to pass in the `current_user` available from the Controller/View context.

I don't have strong feelings about the service approach though. I was mostly following the footprints (and the [service generator](https://github.com/forem/forem/pull/11265)) approach of using the `call` method for initialization. An alternative for this would be to replace the `call` initialization with purely class methods like this:

```ruby
ReCaptcha.call(current_user).enabled? # current approach

# vs

ReCaptcha.enabled?(current_user)
```

Maybe @rhymes, @citizen428 or anyone else have any suggestions/opinions on what would be the best approach here?

Reply:
We do have a service generator (`rails generate service` and [documentation on service objects](https://docs.forem.com/backend/service-objects/).

Maybe in a follow up PR we can split that service objects in two? It looks like you only need the following (coming up with names as I write, definitely open to better ones):

- `ReCaptcha::CheckEnabledGlobally.call(current_user)` (returns true or false, no need to have the negative method)
- `ReCaptcha::CheckEnabledForRegistration.call` (returns true or false, no need to have the negative method)

no need to halt this PR though

Reply:
Thanks for the input @rhymes I'm merging this and sending in a PR right away to refactor the service in two objects. We can discuss the best naming and ways to restructure there 👍🏼